compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java

```diff
@@ -4012,15 +4013,56 @@ public void visit(BLangBinaryExpr binaryExpr) {
         checkDecimalCompatibilityForBinaryArithmeticOverLiteralValues(binaryExpr);
 
         SymbolEnv rhsExprEnv;
-        BType lhsType = checkExpr(binaryExpr.lhsExpr, env);
+        BType lhsType;
+        if (binaryExpr.lhsExpr instanceof BLangNumericLiteral) {
+            switch (binaryExpr.opKind) {
+                case ADD:
+                case SUB:
+                case MUL:
+                case DIV:
+                    if (binaryExpr.expectedType.tag != TypeTags.NONE && binaryExpr.expectedType.tag != TypeTags.ANY) {
+                        lhsType = checkExpectedTypeCompatibility(binaryExpr.lhsExpr,
+                                binaryExpr.expectedType, env);
+                    } else {
+                        lhsType = checkExpr(binaryExpr.lhsExpr, env);;
+                    }
+                    break;
+                default:
+                    lhsType = checkExpr(binaryExpr.lhsExpr, env);
+            }
+        } else {
+            lhsType = checkExpr(binaryExpr.lhsExpr, env);
+        }
+
         if (binaryExpr.opKind == OperatorKind.AND) {
             rhsExprEnv = typeNarrower.evaluateTruth(binaryExpr.lhsExpr, binaryExpr.rhsExpr, env, true);
         } else if (binaryExpr.opKind == OperatorKind.OR) {
             rhsExprEnv = typeNarrower.evaluateFalsity(binaryExpr.lhsExpr, binaryExpr.rhsExpr, env);
         } else {
             rhsExprEnv = env;
         }
-        BType rhsType = checkExpr(binaryExpr.rhsExpr, rhsExprEnv);
+
+        BType rhsType;
+
+        if (binaryExpr.rhsExpr instanceof BLangNumericLiteral) {
+            switch (binaryExpr.opKind) {
+                case ADD:
+                case SUB:
+                case MUL:
+                case DIV:
+                    if (binaryExpr.expectedType.tag != TypeTags.NONE && binaryExpr.expectedType.tag != TypeTags.ANY) {
+                        rhsType = checkExpectedTypeCompatibility(binaryExpr.rhsExpr,
+                                binaryExpr.expectedType, rhsExprEnv);
+                    } else {
+                        rhsType = checkExpectedTypeCompatibility(binaryExpr.rhsExpr, lhsType, rhsExprEnv);
+                    }
+                    break;
+                default:
+                    rhsType = checkExpr(binaryExpr.rhsExpr, rhsExprEnv);
+            }
+        } else {
+            rhsType = checkExpr(binaryExpr.rhsExpr, rhsExprEnv);
+        }
```

Comment:
Isn't this similar to L4017-L4035? Shall we check if we can extract this logic out to a method?

Reply:
I think it can, I will extract the logic out to a method